src/Neo/Network/P2P/Payloads/NotValidBefore.cs

```diff
@@ -0,0 +1,46 @@
+using Neo.IO;
+using Neo.Json;
+using Neo.Persistence;
+using Neo.SmartContract.Native;
+using System.IO;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class NotValidBefore : TransactionAttribute
+    {
+        /// <summary>
+        /// Indicates that the transaction is not valid before this height.
+        /// </summary>
+        public uint Height;
+
+        public override TransactionAttributeType Type => TransactionAttributeType.NotValidBefore;
+
+        public override bool AllowMultiple => false;
+
+        public override int Size => base.Size +
+            sizeof(uint); // Height.
+
+        protected override void DeserializeWithoutType(ref MemoryReader reader)
+        {
+            Height = reader.ReadUInt32();
+        }
+
+        protected override void SerializeWithoutType(BinaryWriter writer)
+        {
+            writer.Write(Height);
+        }
+
+        public override JObject ToJson()
+        {
+            JObject json = base.ToJson();
+            json["height"] = Height;
+            return json;
+        }
+
+        public override bool Verify(DataCache snapshot, Transaction tx)
+        {
+            var block_height = NativeContract.Ledger.CurrentIndex(snapshot);
+            return block_height >= Height;
```

Comment:
It wont wait in the mempool until it's valid

Reply:
That's the desired behaviour, the transaction is not supposed to wait. It's the sender's responsibility to send it at the right time.

Reply:
Then... why we need it? the sender can sign it after the desired Height.

Reply:
#1992:
> It can be used to create some transactions in advance with a guarantee that they won't be accepted until specified block, that property will be used for signature collection service (https://github.com/neo-project/neo/issues/1573).

In #1573 context this allows to have some "hostage" transaction that can be used if the main one is not completed in time (which works especially nice with #1991). But in general, any scenario where you can/need to create a transaction now, but you don't want it to be used until some block. Some offline signing scenarios might benefit from it as well.

Reply:
If the transaction is not valid, it will be included in the know hashes list and it won't be valid until an unknown time

Reply:
It's not supposed to ever appear on the P2P network (unless it's wrapped into P2PNotaryPayload) until it becomes valid. It can't really appear there before NVB height, any node will reject it right away, so I doubt anyone will even try sending it.

Reply:
Maybe it's a different problem, but what will happend if it's in known hashes and now it's valid?

How we will know that it was rejected because previously was rejected?

Reply:
1. Ordinary transactions are not that much different wrt this problem, if node A sends to node B a tx T that for example is lacking some signatures, then it's gonna be the same after signatures are added to T.
2. This actually looks like a C# implementation problem as I can leverage it to prevent transaction dissemination through the network, a malicious node can specifically strip witnesses from a valid tx and send it to all peers. The result will depend on the sequence of events, if there are enough bad nodes and if they're quick enough they can effectively block the tx.